falcon/hooks.py

```diff
@@ -39,68 +38,17 @@
 from falcon.util.sync import _wrap_non_coroutine_unsafe
 
 if TYPE_CHECKING:
+    from typing import Concatenate, ParamSpec
+
     import falcon as wsgi
     from falcon import asgi
     from falcon._typing import AsgiResponderMethod
     from falcon._typing import Resource
     from falcon._typing import Responder
     from falcon._typing import ResponderMethod
 
+    _FN = ParamSpec('_FN')
 
-# TODO: when targeting only 3.10+ these protocol would no longer be needed, since
-# ParamSpec could be used together with Concatenate to use a simple Callable
-# to type the before and after functions. This approach was prototyped in
-# https://github.com/falconry/falcon/pull/2234
-class SyncBeforeFn(Protocol):
-    def __call__(
-        self,
-        req: wsgi.Request,
-        resp: wsgi.Response,
-        resource: Resource,
-        params: Dict[str, Any],
-        *args: Any,
-        **kwargs: Any,
-    ) -> None: ...
-
-
-class AsyncBeforeFn(Protocol):
-    def __call__(
-        self,
-        req: asgi.Request,
-        resp: asgi.Response,
-        resource: Resource,
-        params: Dict[str, Any],
-        *args: Any,
-        **kwargs: Any,
-    ) -> Awaitable[None]: ...
-
-
-BeforeFn = Union[SyncBeforeFn, AsyncBeforeFn]
-
-
-class SyncAfterFn(Protocol):
-    def __call__(
-        self,
-        req: wsgi.Request,
-        resp: wsgi.Response,
-        resource: Resource,
-        *args: Any,
-        **kwargs: Any,
-    ) -> None: ...
-
-
-class AsyncAfterFn(Protocol):
-    def __call__(
-        self,
-        req: asgi.Request,
-        resp: asgi.Response,
-        resource: Resource,
-        *args: Any,
-        **kwargs: Any,
-    ) -> Awaitable[None]: ...
-
-
-AfterFn = Union[SyncAfterFn, AsyncAfterFn]
 _R = TypeVar('_R', bound=Union['Responder', 'Resource'])
```

Comment:
I meant my comment about `Resource` actually applies here even more so.

Reply:
sure, but we don't have a base class for resources, so not sure how else we can express this

Reply:
Yes, I meant I just noticed when I replaced `Resource` with a union of protocols, this logic broke, I was forced to restore it back to `object` to make it work in the current form.